### PR TITLE
Adds specs for solr_q_with_local_params search builder function.

### PR DESCRIPTION
**Pull request by dkinzer**
REF projectblacklight/blacklight#1839

### CODE REVIEW

File: spec/models/blacklight/solr/search_builder_spec.rb [hunk 1]
```diff
@@ -675,4 +675,85 @@
...
+
+    context "with only search field solr_local_parameters present" do
+      let(:blacklight_config) do
+        config = Blacklight::Configuration.new
+        config.add_search_field("test_field", solr_local_parameters: { "bizz" => "buzz" })
+        config
+      end
+
+      it "only adds solr_local_parameters to local params" do
+        expect(solr_parameters[:q]).to eq("{!bizz=buzz}test query")
```
Comment (dkinzer) [inline spec/models/blacklight/solr/search_builder_spec.rb]: Shouldn't this be `{!lucene bizz=buzz}test query`

File: spec/models/blacklight/solr/search_builder_spec.rb [hunk 2]
```diff
@@ -675,4 +675,85 @@
...
+      end
+    end
+
+    context "with neither deftype nor local params" do
+      it "does not add local param values" do
+        expect(solr_parameters[:q]).to eq("{!}test query")
```
Comment (dkinzer) [inline spec/models/blacklight/solr/search_builder_spec.rb]: Shouldn't this be either `{!lucene}test query` or just `test query`

**Comment by magibney** [comment]
Hey @dkinzer (and @barmintor )! Thanks for adopting this issue; it's been a while since I thought about it. The question of what to do with no deftype is a little tricky, because no explicitly specified defType in the request means that it would normally use the default defType as configured in solrconfig.xml (solr-side). We can override deftype without subverting users' intentions when deftype is explicitly specified in query parameters, but when it's not explicitly specified, we can't simply say "ok well now you're using the lucene qparser!". Re: "with neither deftype nor local params", we only need to shuffle around defType and localParams if localParams are specified, so I think this case should result in no modification at all.

Part of the reason I let PR 1839 languish (aside from being otherwise occupied) was because I felt less sure about what the right approach was. There are three options that I can think of to address the SOLR-11501 changes to localParams support:
1. Force Solr to continue to accept the legacy behavior (i.e., "use luceneMatchVersion=7.1.0 or something earlier" -- though this seems to be discouraged)
2. Force blacklight users who rely on localParams to update their blacklight config explicitly to remove localParams dependency (or manually convert to defType=lucene, and incorporate the desired qparser directly in the solr_local_parameters setting).
3. The approach pursued by PR 1839 (and carried along here) of trying to ease the upgrade path for blacklight users by detecting localParams when explicitly set in the blacklight_config, and shuffling around parameters to try to achieve the intended effect. (side note, does this just make an end run around the protections afforded by SOLR-11501, making this option equivalent security-wise to choosing the far easier "option 1", above?)

It's clear that something has to be done to allow many blacklight users to upgrade to newer versions of Solr, but I think I'd like to see a little more exploration of use cases, motivations, implications, etc. before pushing forward with this. @adjam, @jkeck, would you be able to weigh in on this?

Even if we went with option 2 as being the right way to handle this (painful though the upgrade may be) it would still probably require some changes to core blacklight (the only one I can think of is changing handling of solr_local_parameters to incorporate the explicit setting of a local qparser (which I think isn't really supported now?)).

The biggest question I have outstanding at the moment is: how are local params and subqueries used in blacklight? I think the advanced search plugin makes some heavy use of the special `_query_="..."` syntax. Are there use cases for which there is no adequate non-localParams alternative?

**Comment by dkinzer** [comment]
@magibney, yeah I agree that some more conversation is required.  For one I'm not sure many BL users ill actually run into problems:


> SOLR-11501: Starting a query string with local-params {!myparser ...} is used to switch the query parser to another, and is intended for use by Solr system developers, not end users doing searches. To reduce negative side-effects of unintended hack-ability, we've limited the cases that local-params will be parsed to only contexts in which the default parser is "lucene" or "func". So if defType=edismax then q={!myparser ...} won't work. In that example, put the desired query parser into defType. Another example is if deftype=edismax then hl.q={!myparser ...} won't work for the same reason. In that example, either put the desired query parser into hl.qparser or set hl.qparser=lucene. Most users won't run into these cases but some will and must change. If you must have full backwards compatibility, use luceneMatchVersion=7.1.0 or something earlier. 
> (David Smiley)
> SOLR-11501: The edismax parser by default no longer allows subqueries that specify a Solr parser using either local-params, or the older _query_ magic field trick. For example {!prefix f=myfield v=enterp} or _query_:"{!prefix f=myfield v=enterp}" are not supported by default anymore. If you want to allow power-users to do this, set uf=*,_query_ or some other value that includes _query_. If you must have full backwards compatibility, use luceneMatchVersion=7.1.0 or something earlier. 

The issue will come only when a parser is set at defType params, but then differs from that inside of the local params construct.  The old Solr allowed for this dynamic switch of the parser.  But, I don't think any BL user does this.  For one at Temple we set defType to edismax and don't change it dynamically in the local params.  Is anyone doing that in the BL community?

The other concern that I'm having after reading the upgrade blurb is that it's not clear that defaulting to defType=lucene under all cases (which this PR would do) is a good thing.  They seem to suggest about that this is a work around for only some cases not all.

**Comment by magibney** [comment]
I initially encountered this problem not by trying to switch query parsers (e.g., `{!someotherparser p1=x p2=y}`), but simply by trying to set local params on the same parser set in deftype (e.g., `{!p1=x p2=y}`). I'd guess you're right that switching parsers would be a rare BL use case (most common to use edismax) -- but the use of `solr_local_parameters` was all over the place in the config. A sanity check would be nice here -- the upgrade notes you pasted into your comment seem to mention only the parser-switch issue -- but I would never have encountered, let alone pursued, this issue if it hadn't been for the fact that `solr_local_parameters` stopped working for me altogether.

Yeah, just re-confirmed this; although the upgrade notes seem to focus on query parser switching, they do say `limited the cases that local-params will be parsed to only contexts in which the default parser is "lucene" or "func"` -- it really does mean "that local-params will be parsed *at all*, regardless of whether the query parser is switched". 

I'm not sure how `solr_local_parameters` got into our codebase, but a quick glance around at some institutions' public blacklight `catalog_controller`s indicates that we are *far* from being alone in leaning heavily on `solr_local_parameters`. I would guess that institutions that still use `solr_local_parameters` are either still using Solr <= 7.1.0, or have disabled the SOLR-11501 change Solr-side (e.g., using `luceneMatchVersion=7.1.0`).

Anyway, I do think this is something that most BL institutions are going to have to address, one way or another. It looks like Stanford has addressed this issue by dropping `solr_local_parameters` in favor of `solr_parameters`. That's solid, and may be the way to go, but definitely requires changes for upgrading. And might be an issue for some more complex use cases?

**Comment by dkinzer** [comment]
OK @magibney, I think I see what is going on: According to the docs I'm reading:

> If no "type" is specified (either explicitly or implicitly) then the lucene parser is used by default. Thus

> fq={!df=summary}solr rocks

> is equivilent to:

> fq={!type=lucene df=summary}solr rocks

source: https://lucene.apache.org/solr/guide/6_6/local-parameters-in-queries.html

So this means that in the example you gave `{!p1=x p2=y})`, A switch of the parser is happening at the local param level unless your you have set `defType=lucene`, which this PR ends up doing explicitly (https://github.com/projectblacklight/blacklight/pull/2018/files#diff-762950d89973be5eb476b938725b37d0R88).

So the question is, when setting up local params are BL devs using the shorthand `{!key=value}` form, because those will definitely run into issues.

This may also answer the question of whether setting `defType=lucene` is a good idea under all circumstance.  I'm thinking it's not.

**Comment by jrochkind** [comment]
> The issue will come only when a parser is set at defType params, but then differs from that inside of the local params construct. The old Solr allowed for this dynamic switch of the parser. But, I don't think any BL user does this. At Temple U. Library we set defType to edismax and don't change it dynamically in the local params. Is anyone doing that in the BL community

I _think_ advanced search _may_ do this, or something like it -- I don't entirely understand the issue.

**Comment by magibney** [comment]
I'm sorry; I hope I'm faithfully describing a confusing reality, and not introducing the confusion myself!
It's easy enough to test, if you have a running instance of Solr > 7.1.0.

This should clarify things:
Request: `/select?defType=edismax&qf=nonexistent_search&pf=&q={!qf=title_1_search}history&debug=true` -- this request sets explicit top-level params (`solr_params`, in BL parlance) for `defType=edismax`, and sets qf to a nonexistent field name, and unsets pf (for simplicity of generated query).
These parameters should return nothing, because it's only configured to look at a nonexistent field. But let's correct that in the local params! The `{!qf=title_1_search}` attempts to reset `qf` locally to search a real field, that *definitely* should return results for the query "history".
But now let's see how the query is parsed: `+(+DisjunctionMaxQuery(((+nonexistent_search:qf +(nonexistent_search:title_1_search nonexistent_search:"title 1 search") +nonexistent_search:history))))`
Our local params do *not* result in a parser switch (even to lucene, which would have caused the search to fail based on the absence of a `df` param). Rather, the local params are searched as plain text input, and the top-level configured `qf` field is searched for "qf" and "title_1_search" as if these were simple query terms. So we get no docs returned.

**Comment by dkinzer** [comment]
@jrochkind yeah it would seem that blacklight_advanced_search which does make use of `{!' and `{! term ` would run into issues. But I just verified that there are passing PRs that are using solr7.4 https://travis-ci.org/projectblacklight/blacklight_advanced_search/jobs/431862851.

**Comment by magibney** [comment]
Prefer to explicitly update existing blacklight configs to conform to new localParams parsing constraints. See #2020.

**Pull request closed without merging**